lib/router.js

```diff
@@ -1,4 +1,143 @@
-Router.prototype.url = function() {
-  var path = this.path.apply(this, arguments);
-  return Meteor.absoluteUrl(path.replace(/^\//, ''));
-};
+SharedRouter = class {
+  constructor() {
+    this._routes = [];
+    this._routesMap = {};
+
+    // holds onRoute callbacks
+    this._onRouteCallbacks = [];
+  }
+
+  route(pathDef, options, group) {
+    if (!/^\/.*/.test(pathDef)) {
+      const message = "route's path must start with '/'";
+      throw new Error(message);
+    }
+    
+    options = options || {};
+
+    const route = new Route(this, pathDef, options, group);
+
+    this._routes.push(route);
+
+
+    if (options.name) {
+      this._routesMap[options.name] = route;
+    }
+  
+    this._triggerRouteRegister(route);
+
+    return route;
+  }
+
+  // XXX this function needs to be cleaned up if possible by removing `if (this.isServer)`
+  // and `if (this.isClient)` if possible
+  path(pathDef, fields, queryParams) {
+    if (this._routesMap[pathDef]) {
+      pathDef = this._routesMap[pathDef].path;
+    }
+
+    let path = "";
+  
+    // Prefix the path with the router global prefix
+    if (this._basePath) {
+      path += "/" + this._basePath + "/";
+    }
+  
+    fields = fields || {};
+    
+    const regExp = /(:[\w\(\)\\\+\*\.\?]+)+/g;
+
+    path += pathDef.replace(regExp, (key) => {
+      const firstRegexpChar = key.indexOf("(");
+      // get the content behind : and (\\d+/)
+      key = key.substring(1, (firstRegexpChar > 0) ? firstRegexpChar : undefined);
+      // remove +?*
+      key = key.replace(/[\+\*\?]+/g, "");
+  
+      const value = fields[key] || "";
+
+      if (encodeURIComponent) {
+        // this is to allow page js to keep the custom characters as it is
+        // we need to encode 2 times otherwise "/" char does not work properly
+        // So, in that case, when I includes "/" it will think it's a part of the
+        // route. encoding 2times fixes it
+        return encodeURIComponent(encodeURIComponent(value));
+      } else {
+        return value;
+      }
+    });
+  
+    // Replace multiple slashes with single slash
+    path = path.replace(/\/\/+/g, "/");
+  
+    // remove trailing slash
+    // but keep the root slash if it's the only one
+    path = path.match(/^\/{1}$/) ? path : path.replace(/\/$/, "");
+  
+    if (Meteor.isClient) {
+      // explictly asked to add a trailing slash
+      if (this.env.trailingSlash.get() && _.last(path) !== "/") {
```

Comment:
Anyway, I moved the trailingSlash code to the shared code and got rid of the `if (Meteor.isClient)`